From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Abhilash G <abhilashg@nvidia.com>
Date: Thu, 1 Jun 2023 10:21:25 +0000
Subject: [PATCH 080/243] NVIDIA: SAUCE: scsi: ufs: Switch UFS init sequence

Switch UFS init sequence from
LS x1 -> HS x2 to
LS x1 -> HS x1 -> HS x2.
This is a software workaround to avoid symbol errors on RX lanes.
Switch UFS gear to max gear prior to configuring max active lanes.

Bug 3586922
Bug 3621817

Change-Id: I5b224397cdf76075f17a9290f805992d1b5aa983
Signed-off-by: Abhilash G <abhilashg@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/linux-stable/+/2914016
Reviewed-by: Laxman Dewangan <ldewangan@nvidia.com>

diff --git a/drivers/ufs/core/ufshcd.c b/drivers/ufs/core/ufshcd.c
index 504600f1e08c..bf28a3d7b1dd 100644
--- a/drivers/ufs/core/ufshcd.c
+++ b/drivers/ufs/core/ufshcd.c
@@ -4692,8 +4692,6 @@ static int ufshcd_change_power_mode(struct ufs_hba *hba,
 	 * - PA_HSSERIES
 	 */
 	ufshcd_dme_set(hba, UIC_ARG_MIB(PA_RXGEAR), pwr_mode->gear_rx);
-	ufshcd_dme_set(hba, UIC_ARG_MIB(PA_ACTIVERXDATALANES),
-			pwr_mode->lane_rx);
 	if (pwr_mode->pwr_rx == FASTAUTO_MODE ||
 			pwr_mode->pwr_rx == FAST_MODE)
 		ufshcd_dme_set(hba, UIC_ARG_MIB(PA_RXTERMINATION), true);
@@ -4701,8 +4699,6 @@ static int ufshcd_change_power_mode(struct ufs_hba *hba,
 		ufshcd_dme_set(hba, UIC_ARG_MIB(PA_RXTERMINATION), false);
 
 	ufshcd_dme_set(hba, UIC_ARG_MIB(PA_TXGEAR), pwr_mode->gear_tx);
-	ufshcd_dme_set(hba, UIC_ARG_MIB(PA_ACTIVETXDATALANES),
-			pwr_mode->lane_tx);
 	if (pwr_mode->pwr_tx == FASTAUTO_MODE ||
 			pwr_mode->pwr_tx == FAST_MODE)
 		ufshcd_dme_set(hba, UIC_ARG_MIB(PA_TXTERMINATION), true);
@@ -4744,6 +4740,20 @@ static int ufshcd_change_power_mode(struct ufs_hba *hba,
 	if (ret) {
 		dev_err(hba->dev,
 			"%s: power mode change failed %d\n", __func__, ret);
+		return ret;
+	}
+
+	ufshcd_dme_set(hba, UIC_ARG_MIB(PA_ACTIVERXDATALANES),
+			pwr_mode->lane_rx);
+	ufshcd_dme_set(hba, UIC_ARG_MIB(PA_ACTIVETXDATALANES),
+			pwr_mode->lane_tx);
+
+	ret = ufshcd_uic_change_pwr_mode(hba, pwr_mode->pwr_rx << 4
+			| pwr_mode->pwr_tx);
+
+	if (ret) {
+		dev_err(hba->dev,
+			"%s: Lane change failed %d\n", __func__, ret);
 	} else {
 		memcpy(&hba->pwr_info, pwr_mode,
 			sizeof(struct ufs_pa_layer_attr));
-- 
2.43.0

